net/xia/ppal_ether/main.c

```diff
@@ -162,6 +162,296 @@ static int local_dump_ether(struct fib_xid *fxid, struct fib_xid_table *xtbl,
+	struct net_device       *host_interface;
+	struct list_head        interface_common_addr;
+	u8                      ha[ALIGN(MAX_ADDR_LEN, sizeof(long))];
+
+	/* WARNING: @xhm_common is of variable size, and
+	 * MUST be the last member of the struct.
+	 */
+	struct fib_xid          xem_common;
+};
+
+/* Main-FXID functions. */
+static inline struct fib_xid_ether_main *fxid_mether(struct fib_xid *fxid)
+{
+	return likely(fxid)
+		? container_of(fxid, struct fib_xid_ether_main, xem_common)
+		: NULL;
+}
+
+/* Main-FXID + INTERFACE. */
+static void attach_neigh_to_interface(struct fib_xid_ether_main *mether)
+{
+	struct ether_interface *einterface;
+
+	einterface = ether_interface_get(mether->host_interface);
+
+	/*When using list_add_tail_rcu the caller must take whatever precautions
+	* are necessary (such as holding appropriate locks) to avoid racing
+	* with another list-mutation primitive, such as list_add_tail_rcu()
+	* or list_del_rcu(), running on this same list.
+	*/
+	spin_lock(&einterface->interface_lock);
+	list_add_tail_rcu(&einterface->list_interface_common_addr,
+			  &mether->interface_common_addr);
+	spin_unlock(&einterface->interface_lock);
+
+	ether_interface_put(einterface);
+}
+
+/* Main-FXID - INTERFACE.. */
+static void detach_neigh_to_interface(struct fib_xid_ether_main *mether)
+{
+	struct ether_interface *einterface;
+
+	einterface = ether_interface_get(mether->host_interface);
+
+	/*When using list_del_rcu the caller must take whatever precautions are
+	* necessary (such as holding appropriate locks) to avoid racing
+	* with another list-mutation primitive, such as list_add_tail_rcu()
+	* or list_del_rcu(), running on this same list.
+	*/
+	spin_lock(&einterface->interface_lock);
+	list_del_rcu(&mether->interface_common_addr);
+	spin_unlock(&einterface->interface_lock);
+
+	ether_interface_put(einterface);
+}
+
+/* ETHER main table operations. */
+static int main_newroute(struct xip_ppal_ctx *ctx, struct fib_xid_table *xtbl,
+			 struct xia_fib_config *cfg)
+{
+	struct xip_deferred_negdep_flush *dnf;
+	struct fib_xid_ether_main *mether;
+	struct net_device *out_interface;
+	struct fib_xid *cur_fxid;
+	u32 nl_flags, bucket;
+	const char *id;
+	int rc;
+
+	/* Check for errors in cfg. */
+	if (!cfg->xfc_dst || !cfg->xfc_odev || !cfg->xfc_lladdr ||
+	    cfg->xfc_lladdr_len != cfg->xfc_odev->addr_len)
+		return -EINVAL;
+
+	out_interface = cfg->xfc_odev;
+	nl_flags      = cfg->xfc_nlflags;
+	id            = cfg->xfc_dst->xid_id;
+
+	if (!(out_interface->flags & IFF_UP) ||
+	    (out_interface->flags & IFF_LOOPBACK))
+		return -EINVAL;
+
+	cur_fxid = ether_rt_iops->fxid_find_lock(&bucket, xtbl, id);
+	if (cur_fxid) {
+		/* Found a matching fxid. */
+		if (cur_fxid->fx_table_id != XRTABLE_MAIN_INDEX) {
+			rc = -EINVAL;
+			goto unlock_bucket;
+		}
+		mether = fxid_mether(cur_fxid);
+
+		if (!memcmp(mether->ha, cfg->xfc_lladdr, cfg->xfc_lladdr_len) &&
+		    (mether->host_interface == out_interface)) {
```

Comment:
Given that `xfc_dst` is going to be generated based on `xfc_odev` and `xfc_lladdr`, as explained in another comment, this test must be true, so we no longer need it.

You can negate the whole expression and put into a `BUG_ON()` for robustness:
`BUG_ON(memcmp(mether->ha, cfg->xfc_lladdr, cfg->xfc_lladdr_len) ||
	(mether->host_interface != out_interface));`